### PR TITLE
 NetCDF CF Time Iterator for GDAL Source

### PR DESCRIPTION
**Pull request by ChristianBeilschmidt**

### CODE REVIEW

**Comment by ChristianBeilschmidt** [comment]
[4D-NetCDF-Daten laden](https://trello.com/c/v2kpoJbp/534-4d-netcdf-daten-laden)

File: operators/src/source/gdal_source/loading_info.rs [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+}
+
+/// Meta data for 4D `NetCDF` CF datasets
+#[derive(PartialEq, Serialize, Deserialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct GdalMetadataNetCdfCf {
+    pub result_descriptor: RasterResultDescriptor,
+    pub params: GdalDatasetParameters,
+    pub start: TimeInstance,
+    pub end: TimeInstance, // TODO: Use this or time dimension size (number of steps)?
```
Comment (jdroenner) [inline operators/src/source/gdal_source/loading_info.rs:118]: what is the "time dimension size"?

File: operators/src/source/gdal_source/loading_info.rs [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+    pub step: TimeStep,
+    pub band_offset: usize,
```
Comment (jdroenner) [inline operators/src/source/gdal_source/loading_info.rs:120]: what is this? Add a comment how it works

File: operators/src/source/gdal_source/loading_info.rs [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+            .snap_relative(self.start, query.time_interval.start())?;
+
+        let snapped_interval =
+            TimeInterval::new_unchecked(snapped_start, query.time_interval.end()); // TODO: snap end?
+
+        let time_iterator =
+            TimeStepIter::new_with_interval_incl_start(snapped_interval, self.step)?;
+
+        // TODO: REMOVE
+        dbg!(self.band_offset);
```
Comment (jdroenner) [inline operators/src/source/gdal_source/loading_info.rs:139]: remove this

File: operators/src/source/gdal_source/loading_info.rs [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+            time_step_iter,
+            params,
+            step,
+            dataset_time_start,
+            max_t2,
+            band_offset,
+        }
+    }
+
+    // fn band_offset(reference_time: TimeInstance, step: TimeStep, input_time: TimeInstance) -> usize {
```
Comment (jdroenner) [inline operators/src/source/gdal_source/loading_info.rs:253]: will we need this later?

File: operators/src/source/gdal_source/loading_info.rs [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+            0
+        } else {
+            1 + self
+                .step
+                .num_steps_in_interval(TimeInterval::new_unchecked(self.dataset_time_start, t1))
+                .unwrap() // TODO: what to do if this fails?
+        };
+
+        let mut params = self.params.clone();
+        params.rasterband_channel = self.band_offset + 1 + steps_between as usize;
```
Comment (jdroenner) [inline operators/src/source/gdal_source/loading_info.rs:290]: how do we know that the time is ordered  increasing in the netcdf?

File: operators/src/source/gdal_source/loading_info.rs [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+}
+
+impl Iterator for NetCdfCfGdalLoadingInfoPartIterator {
+    type Item = Result<GdalLoadingInfoPart>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let t1 = self.time_step_iter.next()?;
+
+        // our first band is the reference time
+        // TODO: what if it intersects the reference time with t1 < ref?
```
Comment (jdroenner) [inline operators/src/source/gdal_source/loading_info.rs:265]: if it intersects it should also produce data. the reference time is no hard bound.

File: operators/src/source/gdal_source/loading_info.rs [hunk 7]
```diff
@@ -0,0 +1,535 @@
...
+        // snap t1 relative to reference time
+        // TODO: error handling instead of `ok`
+        let t1 = self.step.snap_relative(self.dataset_time_start, t1).ok()?;
+
+        let t2 = t1 + self.step;
+        let t2 = t2.unwrap_or(self.max_t2);
+
+        let time_interval = TimeInterval::new_unchecked(t1, t2);
+
+        // TODO: off by one if date is larger than reference time
```
Comment (jdroenner) [inline operators/src/source/gdal_source/loading_info.rs:279]: can we fix this now?

File: operators/src/source/gdal_source/loading_info.rs [hunk 8]
```diff
@@ -0,0 +1,535 @@
...
+            1 + self
+                .step
+                .num_steps_in_interval(TimeInterval::new_unchecked(self.dataset_time_start, t1))
+                .unwrap() // TODO: what to do if this fails?
+        };
+
+        let mut params = self.params.clone();
+        params.rasterband_channel = self.band_offset + 1 + steps_between as usize;
+
+        // TODO: REMOVE
```
Comment (jdroenner) [inline operators/src/source/gdal_source/loading_info.rs:292]: remove todo

File: operators/src/source/gdal_source/loading_info.rs [hunk 9]
```diff
@@ -0,0 +1,535 @@
...
+
+/// one temporal slice of the dataset that requires reading from exactly one Gdal dataset
+#[derive(Debug, Clone, PartialEq)]
+pub struct GdalLoadingInfoPart {
+    pub time: TimeInterval,
+    pub params: GdalDatasetParameters,
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (jdroenner) [inline operators/src/source/gdal_source/loading_info.rs:342]: We should  check if there are  tests for MEtaData in the "old" file and move them  here.

File: services/src/datasets/external/netcdfcf/error.rs [hunk 10]
```diff
@@ -0,0 +1,92 @@
+use gdal::errors::GdalError;
+use geoengine_datatypes::dataset::DatasetProviderId;
+use snafu::Snafu;
+
+// TODO: `Clone` and `PartialEq`
```
Comment (jdroenner) [inline services/src/datasets/external/netcdfcf/error.rs:5]: why is it needed later?

File: services/src/datasets/external/netcdfcf/mod.rs [hunk 11]
```diff
@@ -0,0 +1,1109 @@
+use std::collections::VecDeque;
+use std::ffi::{c_void, CStr, CString};
+use std::os::raw::{c_char, c_int};
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+
```
Comment (jdroenner) [inline services/src/datasets/external/netcdfcf/mod.rs:6]: if you avoid gaps in the use statements fmt will order and merge them all.

File: services/src/datasets/external/netcdfcf/mod.rs [hunk 12]
```diff
@@ -0,0 +1,1109 @@
...
+    }
+}
+
+pub struct NetCdfCfDataProvider {
+    id: DatasetProviderId,
+    path: PathBuf,
+}
+
+/// TODO: This should be part of the GDAL crate
+/// TODO: free group with `GDALGroupRelease()`
```
Comment (jdroenner) [inline services/src/datasets/external/netcdfcf/mod.rs:89]: i guess this is a leak then?

File: services/src/datasets/external/netcdfcf/mod.rs [hunk 13]
```diff
@@ -0,0 +1,1109 @@
...
+#[derive(Debug)]
+struct MdGroup<'d> {
+    c_group: *mut GDALGroupHS,
+    _dataset: &'d Dataset,
+    name: String,
+}
+
+/// TODO: This should be part of the GDAL crate
+/// TODO: free group with `GDALGroupRelease()`
```
Comment (jdroenner) [inline services/src/datasets/external/netcdfcf/mod.rs:98]: s.o.

File: services/src/datasets/external/netcdfcf/mod.rs [hunk 14]
```diff
@@ -0,0 +1,1109 @@
...
+#[derive(Debug)]
+struct MdArray<'g> {
+    c_mdarray: GDALMDArrayH,
+    _group: &'g MdGroup<'g>,
+}
+
+/// TODO: There are potential memory leaks from the C-API
```
Comment (jdroenner) [inline services/src/datasets/external/netcdfcf/mod.rs:105]: we should not merge code where we introduce memory leaks

File: services/src/datasets/external/netcdfcf/mod.rs [hunk 15]
```diff
@@ -0,0 +1,1109 @@
...
+        Ok(Self {
+            c_group,
+            _dataset: dataset,
+            name: "".to_owned(),
+        })
+    }
+
+    /// Don't put a NULL-byte in the name!!!
+    fn attribute_as_string(&self, name: &str) -> Result<String, GdalError> {
+        let name = CString::new(name).expect("no null-byte in name");
```
Comment (jdroenner) [inline services/src/datasets/external/netcdfcf/mod.rs:173]: you should check that name is not empty

File: services/src/datasets/external/netcdfcf/mod.rs [hunk 16]
```diff
@@ -0,0 +1,1109 @@
...
+
+            Self::_string(GDALAttributeReadAsString(c_attribute))
+        };
+
+        Ok(value)
+    }
+
+    /// Don't put a NULL-byte in the name!!!
+    fn dimension_as_string_array(&self, name: &str) -> Result<Vec<String>, GdalError> {
+        let name = CString::new(name).expect("no null-byte in name");
```
Comment (jdroenner) [inline services/src/datasets/external/netcdfcf/mod.rs:190]: s.o.

File: services/src/datasets/external/netcdfcf/mod.rs [hunk 17]
```diff
@@ -0,0 +1,1109 @@
...
+
+        unsafe {
+            let c_group_names = GDALGroupGetGroupNames(self.c_group, options.as_ptr());
+
+            Self::_string_array(c_group_names)
+        }
+    }
+
+    fn open_group(&self, name: &str) -> Result<Self> {
+        let c_name = CString::new(name).expect("no null-byte in name");
```
Comment (jdroenner) [inline services/src/datasets/external/netcdfcf/mod.rs:263]: s.o.

File: services/src/datasets/external/netcdfcf/mod.rs [hunk 18]
```diff
@@ -0,0 +1,1109 @@
...
+        #[allow(clippy::used_underscore_binding)]
+        Ok(Self {
+            c_group,
+            _dataset: self._dataset,
+            name: name.to_string(),
+        })
+    }
+
+    fn datatype_of_numeric_array(&self, name: &str) -> Result<RasterDataType> {
+        let name = CString::new(name).expect("no null-byte in name");
```
Comment (jdroenner) [inline services/src/datasets/external/netcdfcf/mod.rs:287]: s.o.

File: services/src/datasets/external/netcdfcf/mod.rs [hunk 19]
```diff
@@ -0,0 +1,1109 @@
...
+            }
+
+            MdArray {
+                c_mdarray,
+                _group: self,
+            }
+        })
+    }
+}
+/// TODO: There are potential memory leaks from the C-API
```
Comment (jdroenner) [inline services/src/datasets/external/netcdfcf/mod.rs:340]: we should not merge code where we introduce memory leaks

File: services/src/datasets/external/netcdfcf/mod.rs [hunk 20]
```diff
@@ -0,0 +1,1109 @@
...
+                return Err(NetCdfCf4DProviderError::MustBe4DDataset {
+                    number_of_dimensions,
+                });
+            }
+
+            let dimensions = std::slice::from_raw_parts_mut(c_dimensions, number_of_dimensions);
+
+            DimensionSizes {
+                _entity: GDALDimensionGetSize(dimensions[0]) as usize,
+                time: GDALDimensionGetSize(dimensions[1]) as usize,
```
Comment (jdroenner) [inline services/src/datasets/external/netcdfcf/mod.rs:409]: is hte order fix? only for our case or all CF netcdfs??

File: services/src/datasets/external/netcdfcf/mod.rs [hunk 21]
```diff
@@ -0,0 +1,1109 @@
...
+                lat: GDALDimensionGetSize(dimensions[2]) as usize,
+                lon: GDALDimensionGetSize(dimensions[3]) as usize,
+            }
+        })
+    }
+
+    /// Don't put a NULL-byte in the name!!!
+    fn attribute_as_string(&self, name: &str) -> Result<String, GdalError> {
+        let name = CString::new(name).expect("no null-byte in name");
```
Comment (jdroenner) [inline services/src/datasets/external/netcdfcf/mod.rs:418]: s.o.

File: services/src/datasets/external/netcdfcf/mod.rs [hunk 22]
```diff
@@ -0,0 +1,1109 @@
...
+                    }),
+                    name: format!("{title}: {group_title_path} > {entity_name}"),
+                    description: "".to_owned(), // TODO: where to get from file?
+                    tags: vec![],               // TODO: where to get from file?
+                    source_operator: "GdalSource".to_owned(),
+                    result_descriptor: TypedResultDescriptor::Raster(RasterResultDescriptor {
+                        data_type,
+                        spatial_reference,
+                        measurement: Measurement::Unitless, // TODO: where to get from file?
+                        no_data_value: None, // we don't want to open the dataset at this point. We should get rid of the result descriptor in the listing in general
```
Comment (jdroenner) [inline services/src/datasets/external/netcdfcf/mod.rs:571]: you can add a task for that. I guess we need to rework the Metadata stuff in general

File: services/src/datasets/in_memory.rs [hunk 23]
```diff
@@ -152,6 +155,13 @@ impl HashMapStorable for GdalMetaDataStatic {
     }
 }
 
+impl HashMapStorable for GdalMetadataNetCdfCf {
```
Comment (jdroenner) [inline services/src/datasets/in_memory.rs:158]: we should define one Enum for all the GdalMetadataTypes and use that here. If we have to add all the concrete types here  there is no benefit in using a generic type anyway.

File: services/src/datasets/storage.rs [hunk 24]
```diff
@@ -170,16 +171,17 @@ pub enum MetaDataDefinition {
...
 
 impl MetaDataDefinition {
     pub fn source_operator_type(&self) -> &str {
         match self {
             MetaDataDefinition::MockMetaData(_) => "MockDatasetDataSource",
             MetaDataDefinition::OgrMetaData(_) => "OgrSource",
-            MetaDataDefinition::GdalMetaDataRegular(_) | MetaDataDefinition::GdalStatic(_) => {
-                "GdalSource"
-            }
+            MetaDataDefinition::GdalMetaDataRegular(_)
```
Comment (jdroenner) [inline services/src/datasets/storage.rs:182]: we should wrap them into a new ENUM and  move that to the  gdal_source folder.

**Review by jdroenner — CHANGES_REQUESTED** [review]

File: operators/src/source/gdal_source/loading_info.rs (hunk 6 above)
Comment (ChristianBeilschmidt) [reply operators/src/source/gdal_source/loading_info.rs:265]: What should we return? The data is only partially valid and we cannot specify a band here.

File: operators/src/source/gdal_source/loading_info.rs (hunk 7 above)
Comment (ChristianBeilschmidt) [reply operators/src/source/gdal_source/loading_info.rs:279]: It is actually not a TODO.
The steps_between function works on a single, close-open interval.
We use actually snapped, close-close intervals.

File: operators/src/source/gdal_source/loading_info.rs (hunk 2 above)
Comment (ChristianBeilschmidt) [reply operators/src/source/gdal_source/loading_info.rs:125]: 5692f8d

File: operators/src/source/gdal_source/loading_info.rs (hunk 4 above)
Comment (ChristianBeilschmidt) [reply operators/src/source/gdal_source/loading_info.rs]: 5692f8d

File: operators/src/source/gdal_source/loading_info.rs (hunk 7 above)
Comment (ChristianBeilschmidt) [reply operators/src/source/gdal_source/loading_info.rs]: 5692f8d

File: operators/src/source/gdal_source/loading_info.rs (hunk 9 above)
Comment (ChristianBeilschmidt) [reply operators/src/source/gdal_source/loading_info.rs:333]: 8fc75bc

File: services/src/datasets/external/netcdfcf/error.rs (hunk 10 above)
Comment (ChristianBeilschmidt) [reply services/src/datasets/external/netcdfcf/error.rs:5]: It actually sucks if you cannot do this with errors, especially for testing

File: services/src/datasets/external/netcdfcf/mod.rs (hunk 11 above)
Comment (ChristianBeilschmidt) [reply services/src/datasets/external/netcdfcf/mod.rs:6]: Rust Analyzer seems to group imports by std, crate, and other. Do we need a convention, here? If so, can we force this?

File: services/src/datasets/external/netcdfcf/mod.rs (hunk 12 above)
Comment (ChristianBeilschmidt) [reply services/src/datasets/external/netcdfcf/mod.rs:89]: Yes

File: services/src/datasets/external/netcdfcf/mod.rs (hunk 13 above)
Comment (ChristianBeilschmidt) [reply services/src/datasets/external/netcdfcf/mod.rs:98]: Yes

File: services/src/datasets/external/netcdfcf/mod.rs (hunk 14 above)
Comment (ChristianBeilschmidt) [reply services/src/datasets/external/netcdfcf/mod.rs:105]: I don't know, depends on our strategy on how to go an. We shouldn't keep this implementation of MD.

File: services/src/datasets/external/netcdfcf/mod.rs (hunk 15 above)
Comment (ChristianBeilschmidt) [reply services/src/datasets/external/netcdfcf/mod.rs:173]: Doesn't GDAL just return an error if it is not possible to open it?

File: services/src/datasets/external/netcdfcf/mod.rs (hunk 16 above)
Comment (ChristianBeilschmidt) [reply services/src/datasets/external/netcdfcf/mod.rs:190]: s.o.

File: services/src/datasets/external/netcdfcf/mod.rs (hunk 17 above)
Comment (ChristianBeilschmidt) [reply services/src/datasets/external/netcdfcf/mod.rs:263]: s.o.

File: services/src/datasets/external/netcdfcf/mod.rs (hunk 18 above)
Comment (ChristianBeilschmidt) [reply services/src/datasets/external/netcdfcf/mod.rs:287]: s.o.

File: services/src/datasets/external/netcdfcf/mod.rs (hunk 19 above)
Comment (ChristianBeilschmidt) [reply services/src/datasets/external/netcdfcf/mod.rs:340]: s.o.

File: services/src/datasets/external/netcdfcf/mod.rs (hunk 20 above)
Comment (ChristianBeilschmidt) [reply services/src/datasets/external/netcdfcf/mod.rs:409]: CF enforces this order

File: services/src/datasets/external/netcdfcf/mod.rs (hunk 21 above)
Comment (ChristianBeilschmidt) [reply services/src/datasets/external/netcdfcf/mod.rs:418]: s.o.

File: services/src/datasets/in_memory.rs (hunk 23 above)
Comment (ChristianBeilschmidt) [reply services/src/datasets/in_memory.rs:158]: I guess you are right that this is not an elegant implementation but this is out of scope for this task. You could add a new "idea" into our board if you have something in mind.

File: services/src/datasets/storage.rs (hunk 24 above)
Comment (ChristianBeilschmidt) [reply services/src/datasets/storage.rs:182]: s.o.

additionally, keep in mind that this is a different crate

File: services/src/datasets/external/netcdfcf/mod.rs (hunk 12 above)
Comment (ChristianBeilschmidt) [reply services/src/datasets/external/netcdfcf/mod.rs]: I removed all leaks and checked the tests with ASAN:

`RUSTFLAGS="-Z sanitizer=leak" cargo test --package geoengine-services --lib --all-features --target x86_64-unknown-linux-gnu -- datasets::external::netcdfcf::tests`

e109efe

File: services/src/datasets/external/netcdfcf/mod.rs (hunk 13 above)
Comment (ChristianBeilschmidt) [reply services/src/datasets/external/netcdfcf/mod.rs]: I removed all leaks and checked the tests with ASAN:

`RUSTFLAGS="-Z sanitizer=leak" cargo test --package geoengine-services --lib --all-features --target x86_64-unknown-linux-gnu -- datasets::external::netcdfcf::tests`

e109efe

File: services/src/datasets/external/netcdfcf/mod.rs (hunk 14 above)
Comment (ChristianBeilschmidt) [reply services/src/datasets/external/netcdfcf/mod.rs]: I removed all leaks and checked the tests with ASAN:

`RUSTFLAGS="-Z sanitizer=leak" cargo test --package geoengine-services --lib --all-features --target x86_64-unknown-linux-gnu -- datasets::external::netcdfcf::tests`

e109efe

File: services/src/datasets/external/netcdfcf/mod.rs (hunk 19 above)
Comment (ChristianBeilschmidt) [reply services/src/datasets/external/netcdfcf/mod.rs]: I removed all leaks and checked the tests with ASAN:

`RUSTFLAGS="-Z sanitizer=leak" cargo test --package geoengine-services --lib --all-features --target x86_64-unknown-linux-gnu -- datasets::external::netcdfcf::tests`

e109efe

File: operators/src/source/gdal_source/loading_info.rs (hunk 6 above)
Comment (jdroenner) [reply operators/src/source/gdal_source/loading_info.rs:263]: i thought we generate time intervals here?

File: services/src/datasets/external/netcdfcf/mod.rs (hunk 15 above)
Comment (jdroenner) [reply services/src/datasets/external/netcdfcf/mod.rs:189]: i don't know. you should  check that.

File: services/src/datasets/external/netcdfcf/mod.rs (hunk 11 above)
Comment (jdroenner) [reply services/src/datasets/external/netcdfcf/mod.rs:6]: i guess we should let the analyzer order it

Comment (ChristianBeilschmidt) [reply services/src/datasets/external/netcdfcf/mod.rs:6]: This is how the rust analyzer imports stuff. So should I discard that and order it?

If you for personal preference dislike that, how can we setup the analyzer to do it differently?

File: services/src/datasets/external/netcdfcf/mod.rs (hunk 15 above)
Comment (ChristianBeilschmidt) [reply services/src/datasets/external/netcdfcf/mod.rs:189]: https://gdal.org/api/gdalattribute_cpp.html#_CPPv4NK17GDALIHasAttribute13GetAttributesE12CSLConstList

> Returns    the new attribute, or nullptr if case of error

And we check for `is_null` and return an error.

File: operators/src/source/gdal_source/loading_info.rs (hunk 6 above)
Comment (ChristianBeilschmidt) [reply operators/src/source/gdal_source/loading_info.rs]: daf432b 

New answers and new questions.

File: services/src/datasets/external/netcdfcf/mod.rs (hunk 11 above)
Comment (jdroenner) [reply services/src/datasets/external/netcdfcf/mod.rs:6]: just leave it like that

**Review by jdroenner — APPROVED** [review]

**Comment by jdroenner** [comment]
bors merge

**Pull request merged**